### DISCUSSION TITLE
Discussion: should we change the naming schema of development versions?

### DISCUSSION THREAD

**Discussion by math-GH** [FreshRSS/FreshRSS#5718] [Strategy]
Hi,

right now we have following versions:
- production/stable: `1.21.0`
- development/edge/nightly: `1.21.1-dev`

I see here an issue for the naming of the dev version. When we start to develop the next thing into the stable release, we do not know whether it will become later a bugfix release or a stable release.

Example:
After we release `1.21.0` the development was named `1.21.1-dev`. It looks like that it will become a bugfix with the next version naming `1.21.1`. But it will not. It will be the next `1.22.0`

I propose:
Lets name it what it is. After the release of `1.22.0` the next naming should be `1.22.0-plus-dev` or `1.22.0.edge-dev` or something else. The name should base on what happened before and not what could happen at next step.

What do you think?

**Comment by Alkarex** [comment]
First an introduction: with our current resources, we do not maintain old stable versions, and we do not maintain various branches either. The only way we publish various fixes, including security fixes, is by releasing a new version taken from our edge branch (the edge branch being a rolling release).

So when we publish version 1.22.0, we immediately start the next version, and we do not know yet how long it will take; whether it will be a quick minor version or two, or whether there will be a long stable period until 1.23.0. In order to accommodate for the first possibility, and to keep semver order, the development version becomes 1.22.1-dev. It would be more error prone to start 1.23.0-dev and then come back to 1.22.1 due to an urgent patch, because it would break the order.

> I see here an issue for the naming of the dev version. When we start to develop the next thing into the stable release, we do not know whether it will become later a bugfix release or a stable release.

Yes, this is currently the desired behaviour, as explained.

I am open to other names, as long as it is compatible with https://php.net/function.version-compare (edit: and also compatible with https://semver.org/ )

**Reply by math-GH** [reply]
yes, I agree with you.
My purpose has the same idea. After publishing 1.22.0 there would follow a `1.22.0-plus-dev` instead of `1.22.1-dev` (I did not suggest to name it `1.23.0`)

**Reply by Alkarex** [reply]
My concern is that I have never seen a naming like `1.22.0-plus-dev`, so it does not look usual to me, and I was not able to know whether that means something before or after `1.22.0`.
Normally, `1.22.0-whatever` comes **before** `1.22.0`. I had to lookup https://php.net/function.version-compare to find out the behaviour with `1.22.0-plus-dev`, which indeed comes correctly after `1.22.0` (~~only due to its double dash,~~ which requires users a quite advanced understanding of semver).

```php
php > $versions = ['1.23.0', '1.22.0', '1.22.1', '1.22.0-plus-dev', '1.22.0-dev', '1.22.1-dev'];
php > usort($versions, 'version_compare');
php > print_r($versions);
Array
(
    [0] => 1.22.0-dev
    [1] => 1.22.0
    [2] => 1.22.0-plus-dev
    [3] => 1.22.1-dev
    [4] => 1.22.1
    [5] => 1.23.0
)
```

**Reply by Alkarex** [reply]
Hum, it is not even due to the double dash, there is something else going on, ~~which does not seem to be in the specification~~:

```php
php > $versions = ['1.23.0', '1.22.0', '1.22.1', '1.22.0-plus-dev', '1.22.0-whatever', '1.22.0-plusdev', '1.22.0-dev', '1.22.1-dev'];
php > usort($versions, 'version_compare');
php > print_r($versions);
Array
(
    [0] => 1.22.0-whatever
    [1] => 1.22.0-dev
    [2] => 1.22.0
    [3] => 1.22.0-plus-dev
    [4] => 1.22.0-plusdev
    [5] => 1.22.1-dev
    [6] => 1.22.1
    [7] => 1.23.0
)
```

**Reply by Alkarex** [reply]
Ah, it is due to the `-pl` suffix! I was not familiar with that rule

**Reply by Alkarex** [reply]
Ok, the strange `-plus-dev` behaviour is not standard semver but PHP-specific. Example in Node.js

```js
const semver = require('semver');
const versions = ['1.23.0', '1.22.0', '1.22.1', '1.22.0-plus-dev', '1.22.0-whatever', '1.22.0-plusdev', '1.22.0-dev', '1.22.1-dev'];
versions.sort((a, b) => semver.compare(a, b));
console.log(versions);
```

```js
[
  '1.22.0-dev',
  '1.22.0-plus-dev',
  '1.22.0-plusdev',
  '1.22.0-whatever',
  '1.22.0',
  '1.22.1-dev',
  '1.22.1',
  '1.23.0'
]
```

**Reply by Frenzie** [reply]
> My concern is that I have never seen a naming like 1.22.0-plus-dev, so it does not look usual to me, and I was not able to know whether that means something before or after 1.22.0.

Yes, that's what I meant by the current scheme being clearer. 1.22.0-dev, 1.22.0-alpha, beta, rc, that all means the same thing, that it comes before. 

The `-p`/`-pl` seems to be a PHP curiosity meaning patch level, for putting out a 1.22.0-pl that fits in between 1.22.0 and 1.22.1. Abusing that for what `-dev` is made for seems more confusing to me.

A 1.22.1 can always become a 1.23.0 or possibly even a 2.0.0. I see that as a feature, not a bug. :-)

**Reply by Alkarex** [reply]
Let's stick to what we have for now, but let's keep an eye on even clearer possibilities in the future, especially if it becomes commonly used by other projects.

**Comment by Frenzie** [comment]
I think the current scheme is clearer, especially given how development works. 

But at least for the Docker version perhaps something could be done with the likes of `git show -s --format=format:"%cd" --date=short` and `git describe HEAD` or `git rev-parse --short HEAD`.

Though `git describe` specifically doesn't seem to be terribly useful on `edge`, unless that's a local thing.
```
$ git describe HEAD
1.4.0-3006-g83cf7301
```

**Reply by Alkarex** [reply]
For Docker, see for instance:

```console
$ docker inspect --format='{{index .Config.Labels "org.opencontainers.image.revision"}}' freshrss_dev
edge.83cf7301f46c9609da493a412eaad8d10b09bdda

$ docker inspect --format='{{index .Config.Labels "org.opencontainers.image.revision"}}' freshrss
1.21.0.0fe0ce894cbad09757d719dd4b400b9862c1a12a
```

**Reply by Frenzie** [reply]
Sure, but I thought we were referring to what you see in `about` (which only tangentially relates to the string used for update checking).

![image](https://github.com/FreshRSS/FreshRSS/assets/202757/27ce2af2-9589-4538-a9ec-b720903107fb)

That wouldn't have access to such info from the outside afaik; it'd have to be part of the Dockerfile build process.
